[FEATURE] 로그아웃 기능 구현 #167

jnujh opened the pull request:
## 🚀 작업 내용

카카오 소셜 로그인을 사용하는 JWT 기반 인증 시스템에 안전한 로그아웃 기능을 구현했습니다. 이 기능은 우리 서비스 자체의 로그아웃과 카카오 계정의 로그아웃을 모두 처리하여, 사용자에게 일관되고 안전한 로그아웃 경험을 제공합니다.

**주요 구현 사항:**

* **서비스 로그아웃 API 추가 (`POST /api/auth/logout`)**
    * 사용자의 Access Token을 무효화하기 위한 Redis 블랙리스트 로직을 구현했습니다.
    * 서버(Redis)와 클라이언트(HttpOnly 쿠키) 양쪽에서 Refresh Token을 모두 삭제 및 만료시키는 로직을 구현했습니다.
* **카카오 로그아웃 리다이렉트 API 추가 (`GET /api/auth/kakao/logout`)**
    * 카카오의 '카카오계정과 함께 로그아웃' 흐름을 지원하기 위한 공식적인 도착지(Redirect URI) 엔드포인트를 구현했습니다.
* **인증 필터 로직 강화 (`JwtAuthenticationFilter`)**
    * 서버로 들어오는 모든 요청에 대해, 해당 Access Token이 블랙리스트에 등록되어 있는지 확인하는 검증 절차를 추가하여, 로그아웃된 토큰의 재사용을 원천적으로 차단했습니다.
* **통합 테스트 코드 작성 (`AuthControllerTest`)**
    * 로그아웃 성공, 로그아웃된 토큰 사용 차단 등 다양한 시나리오를 검증하는 통합 테스트를 작성하여 기능의 안정성을 확보했습니다.
* **CI/CD 및 배포 환경 설정**
    * 새로운 로그아웃 기능이 로컬, 테스트(CI), 운영(CD) 모든 환경에서 일관되게 동작하도록 `application.yml`, GitHub Actions 워크플로우, `docker-compose.yml` 등 관련 환경 설정을 모두 최신화했습니다.

---

## ⏱️ 소요 시간

6h

---

## 🤔 고민했던 내용

### 1. JWT의 무상태성(Stateless)과 로그아웃의 딜레마

가장 큰 고민은 **"어떻게 JWT 기반 시스템에서 안전한 로그아웃을 구현할 것인가?"** 였습니다.

> **관련 개념: JWT의 무상태성 (Stateless Nature of JWT)**
>
> * JWT는 한 번 발급되면, 서버가 그 토큰의 상태를 기억하지 않아도 유효 기간이 만료될 때까지 그 자체로 유효합니다. 이는 서버의 확장성을 높여주지만, 특정 토큰을 강제로 "만료"시키는 표준적인 방법을 제공하지 않는다는 단점이 있습니다.
>
> **문제점:**
>
> * 사용자가 '로그아웃' 버튼을 눌러 클라이언트에서 토큰을 삭제하더라도, 만약 그 토큰이 탈취당했다면 해커는 남은 유효 기간 동안 계속해서 서버에 접근할 수 있는 심각한 보안 문제가 발생합니다.

* **선택한 전략: Redis를 이용한 블랙리스트(Blacklist) 방식**
    * **선택 이유:** 이 문제를 해결하기 위해, "무효화된 토큰 목록"을 서버 측에 별도로 저장하는 블랙리스트 전략을 선택했습니다. 사용자가 로그아웃을 요청하면, 아직 유효 기간이 남은 그 Access Token을 Redis에 저장합니다.
    * **구현:** `JwtAuthenticationFilter`가 모든 요청을 검사할 때, 본격적인 토큰 검증 전에 Redis의 블랙리스트에 해당 토큰이 있는지 먼저 확인합니다. 만약 존재한다면, "로그아웃된 토큰"으로 간주하고 즉시 요청을 차단(`401 Unauthorized`)합니다.
    * **최적화:** 블랙리스트에 토큰을 저장할 때, 해당 토큰의 **남은 유효 시간(TTL)**을 동일하게 설정했습니다. 이를 통해 토큰이 어차피 만료되면 블랙리스트에서도 자동으로 삭제되어, 데이터가 불필요하게 쌓이는 것을 방지하고 Redis의 메모리를 효율적으로 관리할 수 있습니다.

### 2. 우리 서비스 로그아웃과 카카오 로그아웃의 연동 순서

두 번째 고민은 **"우리 서비스의 로그아웃과 카카오의 로그아웃을 어떤 순서로 처리해야 하는가?"** 였습니다.

> **관련 개념: API 호출 주체(Caller)와 컨텍스트(Context)**
>
> * API가 누구에 의해 호출되는지, 그리고 그 호출에 어떤 정보(헤더, 바디 등)가 담겨 있는지에 따라 API의 역할이 결정됩니다.
>
> **분석:**
>
> * **우리 서비스 로그아웃 (`POST /api/auth/logout`):** 프론트엔드가 `Authorization` 헤더에 우리 서비스의 Access Token을 담아 **직접 호출**합니다. 이 토큰이 있어야만 우리 백엔드가 "어떤 사용자의 토큰을 무효화할지" 명확히 알 수 있습니다.
> * **카카오 로그아웃 도착지 (`GET /api/auth/kakao/logout`):** **카카오 서버의 지시**에 따라 사용자의 브라우저가 페이지를 이동하는 과정에서 호출됩니다. 이 요청에는 우리 서비스의 토큰이 전혀 포함되어 있지 않으므로, 이 API만으로는 토큰 무효화 작업을 할 수 없습니다.

* **결정한 순서:**
    1.  **먼저, 우리 서비스 로그아웃을 처리한다:** 프론트엔드가 `POST /api/auth/logout`을 호출하여 우리 서비스의 토큰을 모두 무효화합니다.
    2.  **그 다음, 카카오 로그아웃을 진행한다:** 1번 단계가 성공적으로 완료되면, 프론트엔드는 사용자를 카카오 로그아웃 URL로 리다이렉트 시킵니다.
* **선택 이유:** 이 순서를 통해, 카카오 로그아웃 과정에서 사용자가 창을 닫는 등의 예외 상황이 발생하더라도, **우리 서비스에서는 이미 로그아웃 처리가 완료되어 보안을 확보**할 수 있습니다.

---

## 💬 리뷰 중점사항

* `AuthServiceImpl`의 `logout` 메서드에서 Access Token 블랙리스트 등록, Redis의 Refresh Token 삭제, HttpOnly 쿠키 만료 로직이 명확하고 안전하게 구현되었는지 확인 부탁드립니다.
* `JwtAuthenticationFilter`에서 Redis 블랙리스트를 조회하여 로그아웃된 토큰을 차단하는 로직이 다른 토큰 검증 로직보다 **먼저** 수행되는지 확인 부탁드립니다.
* `AuthControllerTest`에 작성된 4가지 로그아웃 시나리오 외에, 추가적으로 검증이 필요하다고 생각되는 엣지 케이스가 있는지 의견 부탁드립니다.

---

## 🔗 관련 이슈

* Close #167 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새 기능**
  * 로그아웃 엔드포인트 추가: POST /api/auth/logout로 토큰 무효화 및 200 응답 제공
  * Kakao 로그아웃 후 메인 페이지로 리다이렉트 추가

* **보안 개선**
  * 로그아웃 시 토큰 블랙리스트 적용으로 로그아웃된 토큰 재사용 차단
  * 인증 필터에 블랙리스트 검사 및 만료 토큰 처리 흐름 개선

* **테스트**
  * 로그아웃 및 재발급 관련 통합 테스트 추가 및 인증 필터 동작 검증 테스트 업데이트
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jnujh committed:
feat: 카카오 및 서비스 로그아웃 엔드포인트 추가 (#167)

카카오 소셜 로그아웃 및 서비스 자체 로그아웃 기능을 구현하기 위한 두 가지 엔드포인트를 AuthController에 추가합니다.

1. 서비스 로그아웃 API 추가 (POST /api/auth/logout)
•클라이언트가 로그아웃을 요청할 때 호출하는 주 API입니다.
•Authorization 헤더의 Access Token을 받아 블랙리스트에 추가하고, HttpOnly 쿠키의 Refresh Token을 무효화하는 로직을 처리합니다.
(실제 로직은 AuthService에서 구현 예정)
•서버의 상태(토큰 무효화)를 변경하는 작업이므로 POST 메서드를 사용합니다.

2. 카카오 로그아웃 리다이렉트 API 추가 (GET /api/auth/kakao/logout)
•카카오의 '카카오계정과 함께 로그아웃' 기능 사용 시, 카카오 서버가 사용자를 리다이렉트시키는 도착지(logout_redirect_uri)입니다.
•별도의 비즈니스 로직 없이, 사용자를 최종적으로 서비스의 메인 페이지로 리다이렉트시키는 '징검다리' 역할을 합니다.
•이 엔드포인트는 로그아웃 흐름의 유연성과 확장성(예: 로깅)을 위해 유지합니다.

이 두 엔드포인트는 전체 로그아웃 흐름을 구성하며, JWT 토큰 무효화와 카카오 계정 세션 만료를 분리하여 처리하는 기반을 마련합니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/auth/controller/AuthController.java

jnujh committed:
feat: 카카오 및 서비스 로그아웃 엔드포인트 추가 (#167)

카카오 소셜 로그아웃 및 서비스 자체 로그아웃 기능을 구현하기 위한 두 가지 엔드포인트를 AuthController에 추가합니다.

1. 서비스 로그아웃 API 추가 (POST /api/auth/logout)
•클라이언트가 로그아웃을 요청할 때 호출하는 주 API입니다.
•Authorization 헤더의 Access Token을 받아 블랙리스트에 추가하고, HttpOnly 쿠키의 Refresh Token을 무효화하는 로직을 처리합니다.
(실제 로직은 AuthService에서 구현 예정)
•서버의 상태(토큰 무효화)를 변경하는 작업이므로 POST 메서드를 사용합니다.

2. 카카오 로그아웃 리다이렉트 API 추가 (GET /api/auth/kakao/logout)
•카카오의 '카카오계정과 함께 로그아웃' 기능 사용 시, 카카오 서버가 사용자를 리다이렉트시키는 도착지(logout_redirect_uri)입니다.
•별도의 비즈니스 로직 없이, 사용자를 최종적으로 서비스의 메인 페이지로 리다이렉트시키는 '징검다리' 역할을 합니다.
•이 엔드포인트는 로그아웃 흐름의 유연성과 확장성(예: 로깅)을 위해 유지합니다.

이 두 엔드포인트는 전체 로그아웃 흐름을 구성하며, JWT 토큰 무효화와 카카오 계정 세션 만료를 분리하여 처리하는 기반을 마련합니다.

jnujh committed:
feat: AuthService logout 메서드 구현 (#167)

AuthService 인터페이스에 선언된 logout 메서드의 실제 로직을 구현합니다.
이 메서드는 JWT 기반 인증 시스템에서 안전한 로그아웃을 위해 다음 세 가지 핵심 작업을 수행합니다.

1.Access Token 블랙리스트 등록:
•RedisTemplate을 사용하여 전달받은 Access Token을 Redis에 블랙리스트로 등록합니다.
•이때, 토큰의 남은 유효 시간을 계산하여 TTL(Time To Live)로 설정함으로써,
토큰이 자연 만료되면 블랙리스트에서도 자동으로 삭제되도록 하여 데이터의 효율적인 관리를 도모합니다.

2.Refresh Token 서버 측 삭제:
•Access Token에서 추출한 사용자 ID를 기반으로, RefreshTokenRepository를 통해
Redis에 저장된 해당 사용자의 Refresh Token을 완전히 삭제합니다.
이를 통해 새로운 Access Token 발급을 원천적으로 차단합니다.

3.Refresh Token 클라이언트 측 삭제:
•HttpServletResponse에 Max-Age=0으로 설정된 동일한 이름의 쿠키를 추가하여 응답합니다.
•브라우저는 이 응답을 받고 기존의 HttpOnly Refresh Token 쿠키를 덮어쓰고 즉시 삭제하게 됩니다.

이 세 가지 로직을 통해, 서버와 클라이언트 양쪽에서 토큰을 모두 무효화하여 완전한 로그아웃을 구현합니다

jnujh committed:
feat: AuthService logout 메서드 구현 (#167)

AuthService 인터페이스에 선언된 logout 메서드의 실제 로직을 구현합니다.
이 메서드는 JWT 기반 인증 시스템에서 안전한 로그아웃을 위해 다음 세 가지 핵심 작업을 수행합니다.

1.Access Token 블랙리스트 등록:
•RedisTemplate을 사용하여 전달받은 Access Token을 Redis에 블랙리스트로 등록합니다.
•이때, 토큰의 남은 유효 시간을 계산하여 TTL(Time To Live)로 설정함으로써,
토큰이 자연 만료되면 블랙리스트에서도 자동으로 삭제되도록 하여 데이터의 효율적인 관리를 도모합니다.

2.Refresh Token 서버 측 삭제:
•Access Token에서 추출한 사용자 ID를 기반으로, RefreshTokenRepository를 통해
Redis에 저장된 해당 사용자의 Refresh Token을 완전히 삭제합니다.
이를 통해 새로운 Access Token 발급을 원천적으로 차단합니다.

3.Refresh Token 클라이언트 측 삭제:
•HttpServletResponse에 Max-Age=0으로 설정된 동일한 이름의 쿠키를 추가하여 응답합니다.
•브라우저는 이 응답을 받고 기존의 HttpOnly Refresh Token 쿠키를 덮어쓰고 즉시 삭제하게 됩니다.

이 세 가지 로직을 통해, 서버와 클라이언트 양쪽에서 토큰을 모두 무효화하여 완전한 로그아웃을 구현합니다

jnujh committed:
feat: 로그아웃된 토큰 에러 코드 추가 (#167)

ErrorCode 열거형에 BLACKLISTED_TOKEN을 새로 추가합니다.
이 에러 코드는 로그아웃 처리되어 Redis 블랙리스트에 등록된 Access Token으로 API 요청이 들어왔을 때, 이를 명시적으로 처리하기 위해 정의되었습니다.
향후 JwtAuthenticationFilter에서 블랙리스트에 등록된 토큰을 감지하면,
InvalidJwtException(ErrorCode.BLACKLISTED_TOKEN)을 발생시키게 됩니다.
이를 통해 GlobalExceptionHandler는 해당 예외를 일관되게 처리하여, 클라이언트에게 명확한 401 Unauthorized 응답을 보낼 수 있습니다.

jnujh committed:
refactor: 전화번호 형식 변경 테스트 적용

jnujh committed:
feat: JWT 필터에 토큰 블랙리스트 검증 로직 추가 (#167)

JwtAuthenticationFilter에 Redis를 조회하여 Access Token이 블랙리스트에 등록되어 있는지 확인하는 로직을 추가합니다.
이전 단계에서 AuthService에 로그아웃된 토큰을 Redis 블랙리스트에 등록하는 로직을 구현했지만,
실제 요청을 가로채는 필터단에서 이 블랙리스트를 확인하지 않으면 로그아웃이 실질적인 효과를 갖지 못합니다.

주요 변경 사항:
•RedisTemplate을 의존성으로 주입받아 Redis에 접근할 수 있도록 수정합니다.
•doFilterInternal 메서드에서 토큰 유효성을 검증하기 전에, 먼저 Redis에 blacklist:{accessToken} 키가 존재하는지 확인합니다.
•만약 키가 존재한다면, 이는 로그아웃 처리된 토큰으로 간주하고
InvalidJwtException(ErrorCode.BLACKLISTED_TOKEN)을 발생시켜 요청을 즉시 차단합니다.

이로써, 유효 기간이 남아있더라도 로그아웃된 Access Token을 사용한 모든 API 요청을 원천적으로 차단하여 보안을 강화합니다

jnujh committed:
test: 로그아웃 기능 통합 테스트 코드 작성 (#167)

로그아웃 기능의 정확성과 안정성을 검증하기 위해 AuthControllerTest 통합 테스트 클래스를 추가합니다.

주요 테스트 시나리오:
1.정상 로그아웃: 유효한 토큰으로 로그아웃 요청 시 200 OK를 반환하는지 검증합니다.
2.블랙리스트 토큰 접근 차단: 로그아웃 처리된 Access Token으로 보호된 API에 접근 시 401 Unauthorized를 반환하는지 검증합니다.
3.삭제된 Refresh Token 재발급 차단: 로그아웃으로 인해 삭제된 Refresh Token으로 재발급 요청 시 401 Unauthorized를 반환하는지 검증합니다.
4.중복 로그아웃 차단: 이미 블랙리스트에 등록된 토큰으로 다시 로그아웃을 시도할 경우 401 Unauthorized를 반환하는지 검증합니다.

jnujh committed:
test: 로그아웃 기능 통합 테스트 코드 작성 (#167)

로그아웃 기능의 정확성과 안정성을 검증하기 위해 AuthControllerTest 통합 테스트 클래스를 추가합니다.

주요 테스트 시나리오:
1.정상 로그아웃: 유효한 토큰으로 로그아웃 요청 시 200 OK를 반환하는지 검증합니다.
2.블랙리스트 토큰 접근 차단: 로그아웃 처리된 Access Token으로 보호된 API에 접근 시 401 Unauthorized를 반환하는지 검증합니다.
3.삭제된 Refresh Token 재발급 차단: 로그아웃으로 인해 삭제된 Refresh Token으로 재발급 요청 시 401 Unauthorized를 반환하는지 검증합니다.
4.중복 로그아웃 차단: 이미 블랙리스트에 등록된 토큰으로 다시 로그아웃을 시도할 경우 401 Unauthorized를 반환하는지 검증합니다.

jnujh committed:
refactor: 주석 제거

jnujh committed:
fix: HTTP method 변경 (#167)

jnujh committed:
chore: CI/CD 및 배포 환경에 로그아웃 설정 적용 (#167)

새롭게 구현된 로그아웃 기능을 CI(지속적 통합) 및 CD(지속적 배포) 환경에서 올바르게 동작시키기 위해,
관련 워크플로우와 Docker 설정 파일을 수정합니다.

주요 변경 사항:
1.ci-on-pr.yml 수정 (CI 환경 설정)
•env 섹션에 KAKAO_LOGOUT_REDIRECT_URI_TEST와 FRONTEND_MAIN_PAGE_URI_TEST를 추가하여,
GitHub Secrets로부터 테스트용 URI 값을 읽어오도록 설정했습니다.
•cat 명령어로 application-test.yml을 동적으로 생성하는 스크립트에, 위 환경 변수들을 사용하는
logout-redirect-uri와 frontend.main-page-uri 프로퍼티를 추가했습니다.

2.docker-compose.yml 수정 (Docker 컨테이너 환경 정의)
•app 서비스의 environment 섹션에 KAKAO_LOGOUT_REDIRECT_URI_PROD와 FRONTEND_MAIN_PAGE_URI_PROD를 추가했습니다.

3.deploy-with-docker.yml 수정 (CD 환경 설정)
•EC2 인스턴스에 .env 파일을 생성하는 printf 명령어의 형식 문자열과 변수 목록을 수정했습니다.
•KAKAO_LOGOUT_REDIRECT_URI_PROD와 FRONTEND_MAIN_PAGE_URI_PROD라는 이름의 GitHub Secrets 값을 읽어와
.env 파일에 기록하도록 추가했습니다.

jnujh committed:
chore: CI/CD 및 배포 환경에 로그아웃 설정 적용 (#167)

새롭게 구현된 로그아웃 기능을 CI(지속적 통합) 및 CD(지속적 배포) 환경에서 올바르게 동작시키기 위해,
관련 워크플로우와 Docker 설정 파일을 수정합니다.

주요 변경 사항:
1.ci-on-pr.yml 수정 (CI 환경 설정)
•env 섹션에 KAKAO_LOGOUT_REDIRECT_URI_TEST와 FRONTEND_MAIN_PAGE_URI_TEST를 추가하여,
GitHub Secrets로부터 테스트용 URI 값을 읽어오도록 설정했습니다.
•cat 명령어로 application-test.yml을 동적으로 생성하는 스크립트에, 위 환경 변수들을 사용하는
logout-redirect-uri와 frontend.main-page-uri 프로퍼티를 추가했습니다.

2.docker-compose.yml 수정 (Docker 컨테이너 환경 정의)
•app 서비스의 environment 섹션에 KAKAO_LOGOUT_REDIRECT_URI_PROD와 FRONTEND_MAIN_PAGE_URI_PROD를 추가했습니다.

3.deploy-with-docker.yml 수정 (CD 환경 설정)
•EC2 인스턴스에 .env 파일을 생성하는 printf 명령어의 형식 문자열과 변수 목록을 수정했습니다.
•KAKAO_LOGOUT_REDIRECT_URI_PROD와 FRONTEND_MAIN_PAGE_URI_PROD라는 이름의 GitHub Secrets 값을 읽어와
.env 파일에 기록하도록 추가했습니다.

ji-mim commented on src/test/java/com/kakaotech/team18/backend_server/domain/auth/controller/AuthControllerTest.java:
```diff
@@ -0,0 +1,142 @@
+        // 2. 로그아웃에 사용했던 Access Token으로 보호된 API에 접근을 시도합니다.
+        ResultActions resultActions = mockMvc.perform(
+                get("/api/auth/reissue")
+                        .header("Authorization", "Bearer " + accessToken)
```
reissue API에 accessToken을 보내는 것 자체에서 에러가 발생해야 하는거죠?

jnujh replied:
넵 맞습니다.
추가로 "블랙리스트에 등록된 토큰으로 API 요청을 하면 에러가 발생한다." 이 의미도 있습니다 !

ji-mim commented on docker-compose.yml:
```diff
@@ -65,6 +65,8 @@ services:
       KAKAO_REDIRECT_URI_PROD: ${KAKAO_REDIRECT_URI_PROD}
+      KAKAO_LOGOUT_REDIRECT_URI_PROD: ${KAKAO_LOGOUT_REDIRECT_URI_PROD}
+      FRONTEND_MAIN_PAGE_URI_PROD: ${FRONTEND_MAIN_PAGE_URI_PROD}
       SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_KAKAO_CLIENT_ID: ${KAKAO_CLIENT_ID}
```
이 환경변수들은 혹시 어디서 사용될까요? 컨트롤러 쪽에서 사용하는건 아닌 것 같아서요

jnujh replied:
KAKAO_LOGOUT_REDIRECT_URI_PROD는 카카오 OAuth2 서비스에 등록된 "로그아웃 후 리다이렉트될 URI"입니다. 
사용자가 카카오 계정에서 로그아웃을 완료했을 때, 카카오 서버가 이 URI로 사용자의 브라우저를 다시 우리 서비스로 돌려보내도록 설정하는 역할을 합니다,

RONTEND_MAIN_PAGE_URI_PROD는 우리 서비스의 프론트엔드 애플리케이션 메인 페이지의 URI를 나타냅니다. 
백엔드에서 로그아웃 완료 후에 사용자를 프론트엔드의 특정 페이지로 리다이렉트해야 할 때 사용됩니다.

ji-mim reviewed (approved):
Redis에 대해서 잘 모르는데 Map 처럼 key 값으로 값을 지정해두고 블랙리스트처럼 관리할 수 있는게 신기하네요! 나중에 여유 있을 때 같이 로그아웃시 또 어떤 방법을 사용할 수 있을지 찾아보고 비교해보는 것도 좋을 것 같아요.

새벽에 수고 많으셨어요!

jnujh committed:
feat: 토큰 만료 에러 코드 세분화 (#172)

Access Token 만료와 Refresh Token 만료를 구분하기 위해 에러 코드를 분리합니다.
•Access Token 만료: EXPIRED_ACCESS_TOKEN
•Refresh Token 만료: EXPIRED_REFRESH_TOKEN

이를 통해 클라이언트에서 각 에러에 따른 분기 처리가 가능해집니다.

jnujh committed:
Feat: Refresh Token 만료 시 전용 에러 코드 반환 (#172)

기존에는 Access Token과 Refresh Token 만료 시 동일한 예외가 발생하여 클라이언트에서 두 상황을 구분하기 어려웠습니다.

토큰 재발급(`reissue`) 로직에서 Refresh Token 검증 중 `ExpiredJwtException`이 발생할 경우,
이를 `CustomException(ErrorCode.EXPIRED_REFRESH_TOKEN)`으로 전환하여 던지도록 수정했습니다.

이를 통해 클라이언트는 Refresh Token이 만료되었음을 명확히 인지하고, 사용자에게 재로그인을 안내하는 등의 후속 조치를 할 수 있습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/auth/service/AuthServiceImpl.java

jnujh committed:
Refactor: Access Token 만료 시 EXPIRED_ACCESS_TOKEN 에러 반환 (#172)

토큰 만료 에러 코드를 세분화하는 작업의 일환으로, `JwtAuthenticationFilter`를 수정했습니다.

요청의 Access Token이 만료되어 `ExpiredJwtException`이 발생했을 때,
기존의 범용적인 에러 코드 대신 명시적인 `ErrorCode.EXPIRED_ACCESS_TOKEN`을 사용하도록 변경했습니다.

이 변경으로 클라이언트는 Access Token 만료 상황을 명확하게 식별하고, 토큰 재발급 플로우를 트리거할 수 있습니다.

jnujh committed:
Test: Refresh Token 만료 시 재발급 실패 테스트 추가 (#172)

토큰 재발급 API(`/api/auth/reissue`) 호출 시, Refresh Token이 만료된 경우에 대한 컨트롤러 테스트 케이스를 추가합니다.

AuthService가 `CustomException(ErrorCode.EXPIRED_REFRESH_TOKEN)`을 던지는 상황을 모킹하고,
API 호출 결과가 예상대로 HTTP 401 상태 코드와 `EXPIRED_REFRESH_TOKEN` 에러 코드를 포함하는지 검증합니다.

jnujh committed:
Test: Access Token 만료 에러 코드 변경에 따른 테스트 수정 (#172)

`JwtAuthenticationFilter`에서 Access Token 만료 시 반환하는 에러 코드가 `EXPIRED_ACCESS_TOKEN`으로 변경됨에 따라, 관련 통합 테스트 코드를 수정합니다.

만료된 토큰으로 API 요청 시, 응답 JSON의 `message` 필드가
`ErrorCode.EXPIRED_ACCESS_TOKEN`에 정의된 메시지와 일치하는지 검증하도록 `jsonPath` 단언문을 업데이트했습니다.

jnujh committed:
Refactor: JwtAuthenticationFilter의 자동 등록 방지

기존에는 @Component 어노테이션으로 인해 JwtAuthenticationFilter가 Spring의 컴포넌트 스캔 과정에서 자동으로 Bean으로 등록되었습니다.
이로 인해 SecurityConfig의 ignoring() 설정과 무관하게 필터가 동작하여, 필터가 이중으로 등록되는 문제가 발생했습니다.

이 문제를 해결하고 필터의 생성 및 등록 시점을 명확하게 제어하기 위해 @Component 어노테이션을 제거합니다.

이제 JwtAuthenticationFilter는 SecurityConfig 내에서 수동으로 생성되고 필터 체인에 등록되어,
정확히 한 번만, 그리고 의도된 보안 규칙에 따라서만 동작하게 됩니다.

jnujh committed:
Fix: 필터 수동 등록 및 ignoring() 설정으로 reissue 무한 루프 해결

토큰 재발급(/api/auth/reissue) 요청 시, 만료된 Access Token으로 인해 인증 필터에서 차단되어 무한 루프가 발생하는 문제를 해결합니다.

1.  WebSecurityCustomizer 추가:
    -   /api/auth/** 등 인증이 필요 없는 경로를 Spring Security 필터 체인에서 완전히 제외(ignore)하도록 설정했습니다.
    -   이를 통해 /api/auth/reissue 요청이 헤더에 만료된 토큰을 포함하더라도, 인증 필터를 거치지 않고 컨트롤러까지 안전하게 도달할 수 있습니다.

2.  JwtAuthenticationFilter 수동 등록:
    - 기존의 필터 이중 등록 문제를 해결하기 위해, SecurityConfig가 JwtAuthenticationFilter를 직접 생성(new)하고 필터 체인에 등록하도록 변경했습니다.
    - 이를 위해 생성자를 통해 필요한 의존성(JwtProvider 등)을 주입받도록 구조를 개선했습니다.

jnujh committed:
Test: /api/auth/** 경로에 대한 인증 필터 무시 검증 테스트 추가

SecurityConfig에 추가된 WebSecurityCustomizer의 ignoring() 설정이 올바르게 동작하는지 검증하는 통합 테스트 케이스를 추가합니다.

테스트 시나리오는 다음과 같습니다.
- 일부러 만료된 Access Token을 헤더에 담아 /api/auth/reissue 경로로 요청합니다.
- 인증 필터가 무시되었다면 요청은 401(Unauthorized)이 아닌, 컨트롤러 레벨의 에러(400 Bad Request)를 반환해야 합니다.

gary5876 commented on src/main/java/com/kakaotech/team18/backend_server/domain/auth/controller/AuthController.java:
```diff
@@ -71,4 +73,31 @@ public ResponseEntity<ReissueResponseDto> reissue(@RequestHeader("Authorization"
+    public String kakaoLogout() {
+
+        // 최종적으로 프론트엔드의 메인 페이지로 리다이렉트
+        return "redirect:https://www.dongarium.co.kr/";
```
제가 이해를 잘 못한건지 모르겠는데, 
이 부분을 보고 백에서 리디이렉트까지 시켜줘야하나? 라는 의문이 생겼어요. 카카오 로그인밖에 없는 상황에 카카오 로그아웃과 그냥 로그아웃이 분리되어야하나? 하는 의문도 있네요

jnujh replied:
그냥 로그아웃 (/api/auth/logout): 우리 서비스에서만 로그아웃합니다. 즉, 우리 서비스의 JWT 토큰만 무효화됩니다. 
사용자는 여전히 카카오에는 로그인된 상태입니다.

카카오계정과 함께 로그아웃: 우리 서비스와 카카오 양쪽 모두에서 로그아웃합니다. 
사용자가 "완전한 로그아웃"을 원할 때 이 기능을 제공해야 합니다. 
그렇지 않으면, 우리 서비스에서 로그아웃해도 다시 로그인 버튼을 누르면 카카오 로그인 화면 없이 바로 로그인될 수 있기 때문입니다.

이러한 이유로 카카오 로그아웃과 그냥 로그아웃을 분리했습니다 !

---

그리고 백에서 리다이렉트 시키도록 한 이유는 
카카오 로그아웃 완료 후 추가적인 처리 -> 특히 로깅 !! 등을 적용하려고 했는데 
제가 깜빡했네요 ㅎ
추가하겠습니다

jnujh replied:
백을 거치지 않고 곧바로 프론트엔드로 리다이렉트 시키는게 가능할 거 같긴 합니다.

백을 거치고 로깅 적용 vs 곧바로 프론트로 리다이렉트 
어떤게 더 나을까요 ? ?

ji-mim replied:
혹시 유저가 페이지를 로그인 상태로 저희 서비스에서 이탈하거나 해서 토큰이 만료되면 로그아웃 되기도 할까요? 저는 이대로 가고 로그 추가해도 좋을 것 같아요~

jnujh committed:
refactor: h2콘솔 web.ignoring 추가

jnujh committed:
refactor: RefreshToken 만료에 대핸 구체적인 예외 클래스 생성

jnujh committed:
Merge branch 'develop' into feature/logout#167

jnujh committed:
fix: JWT 필터가 토큰 재발급 요청을 차단하는 문제 해결

## 문제 원인

만료된 액세스 토큰을 헤더에 담아 토큰 재발급 API(`/api/auth/reissue`)를 호출할 때,
`JwtAuthenticationFilter`가 컨트롤러에 도달하기 전에 만료된 토큰을 먼저 검증하여
`ExpiredJwtException`을 발생시키는 문제가 있었습니다.

이로 인해 클라이언트는 정상적으로 토큰을 재발급받지 못하고 인증 오류(401 Unauthorized)를 응답받게 되었습니다.

기존에는 이 문제를 해결하기 위해 `SecurityConfig`에서 `ignoring()` 설정을 사용하여
`/api/auth/**` 경로를 Spring Security 필터 체인에서 완전히 제외했습니다.
하지만 이 방식은 해당 경로에 CORS, 보안 헤더 등 다른 보안 설정들이 적용되지 않는 부작용이 있었습니다.

## 해결 방안

이 문제를 근본적으로 해결하고 보안 일관성을 유지하기 위해 다음과 같이 수정했습니다.

1.  **`SecurityConfig.java` 수정**
    *   `ignoring()` 목록에서 `/api/auth/**` 경로를 제거했습니다.
    *   이제 모든 API 요청이 Spring Security 필터 체인을 통과하게 되어, 모든 경로에 일관된 보안 정책(CORS, 보안 헤더 등)이 적용됩니다.

2.  **`JwtAuthenticationFilter.java` 수정**
    *   `doFilterInternal` 메서드 시작 부분에 요청 URI를 확인하는 로직을 추가했습니다.
    *   요청 URI가 `/api/auth/`로 시작하는 경우, 토큰 검증 로직을 건너뛰고 즉시 다음 필터로 요청을 전달하도록 수정했습니다.
    *   이를 통해 토큰 재발급 요청이 만료된 토큰의 영향을 받지 않고 안전하게 컨트롤러에 도달할 수 있게 되었습니다.

jnujh committed:
fix: develop rebase에 맞게 수정

jnujh committed:
fix: 만료된 토큰만 필터에서 예외 처리하도록 수정

기존에는 JwtAuthenticationFilter에서 "/api/auth/**"로 시작하는 모든 요청을 토큰 검증 없이 통과시켰습니다.
이로 인해 Access Token이 만료된 상태에서 토큰 재발급(/reissue)을 요청하는 것은 가능했지만,
유효한 토큰으로 로그아웃(/logout)을 요청했을 때 해당 토큰이 블랙리스트에 있는지 검사하는 로직 또한 건너뛰는 문제가 발생했습니다.
결과적으로 이미 로그아웃된 토큰으로 다시 로그아웃을 시도해도 성공(200 OK)하는 버그가 있었습니다.

이 문제를 해결하기 위해, 필터에서 특정 URL 경로를 기준으로 예외 처리하는 로직을 제거했습니다.
대신, 토큰 검증 과정에서 `ExpiredJwtException`이 발생한 경우에만 요청을 통과시키도록 수정했습니다.

이 변경을 통해 다음과 같은 효과를 기대할 수 있습니다.
- 만료된 Access Token을 가진 재발급 요청은 정상적으로 컨트롤러에 도달합니다.
- 로그아웃 요청 시에는 토큰이 유효한지, 블랙리스트에 등록되어 있는지 필터단에서 올바르게 검증됩니다.

jnujh committed:
fix: 수정사항에 맞게 테스트코드 수정

jnujh committed:
fix: 수정사항에 맞게 테스트코드 수정

jnujh committed:
Merge branch 'develop' into feature/logout#167

jnujh committed:
fix: rebase 과정에서 꼬인 부분 수정

jnujh committed:
fix: 수정사항에 맞게 테스트 수정
